tests/import_source/import_source.cpp

```diff
@@ -0,0 +1,401 @@
+
+    // Add the other direction, expect false as it's already there.
+    EXPECT_FALSE(imp->addComponent(component1));
+
+    // Add the other component so it's an import.
+    component2->setImportSource(imp);
+    EXPECT_TRUE(component2->isImport());
+    EXPECT_EQ(size_t(2), imp->componentCount());
+    EXPECT_EQ(component2, imp->component(1));
+
+    // Remove a component from the import source by index.
+    EXPECT_TRUE(imp->removeComponent(0));
+    EXPECT_EQ(size_t(1), imp->componentCount());
+    EXPECT_FALSE(component1->isImport());
+
+    // Remove component by pointer.
+    EXPECT_TRUE(imp->removeComponent(component2));
+    EXPECT_FALSE(component2->isImport());
+    EXPECT_EQ(size_t(0), imp->componentCount());
+
+    // Remove a component that doesn't exist.
+    EXPECT_FALSE(imp->removeComponent(99));
+    EXPECT_FALSE(imp->removeComponent(component2));
+    libcellml::ComponentPtr nullComponent = nullptr;
+    EXPECT_FALSE(imp->removeComponent(nullComponent));
+
+    // Set up units so it's an import.
+    units1->setImportSource(imp);
+
+    EXPECT_TRUE(units1->isImport());
+    EXPECT_EQ(size_t(1), imp->unitsCount());
+    EXPECT_EQ(units1, imp->units(0));
+    EXPECT_EQ(nullptr, imp->units(99));
+
+    // Add the other direction, expect false as it's already there.
+    EXPECT_FALSE(imp->addUnits(units1));
+
+    // Add the other units so it's an import.
+    units2->setImportSource(imp);
+
+    EXPECT_TRUE(units2->isImport());
+    EXPECT_EQ(size_t(2), imp->unitsCount());
+    EXPECT_EQ(units2, imp->units(1));
+
+    // Remove a units from the import source by index.
+    EXPECT_TRUE(imp->removeUnits(0));
+    EXPECT_EQ(size_t(1), imp->unitsCount());
+    EXPECT_FALSE(units1->isImport());
+
+    // Remove units by pointer.
+    EXPECT_TRUE(imp->removeUnits(units2));
+    EXPECT_FALSE(units2->isImport());
+
+    // Remove a units that doesn't exist.
+    EXPECT_FALSE(imp->removeUnits(99));
+    EXPECT_FALSE(imp->removeUnits(units2));
+    libcellml::UnitsPtr nullUnits = nullptr;
+    EXPECT_FALSE(imp->removeUnits(nullUnits));
+
+    EXPECT_TRUE(model->removeImportSource(0));
+    EXPECT_FALSE(model->removeImportSource(0));
+    model->addImportSource(imp);
+    EXPECT_TRUE(model->removeImportSource(imp));
+    EXPECT_FALSE(model->removeImportSource(imp));
+}
+
+TEST(ImportSource, importSourceMove)
+{
+    auto imp1 = libcellml::ImportSource::create();
+    auto imp2 = libcellml::ImportSource::create();
+
+    auto component = libcellml::Component::create("component");
+    auto units = libcellml::Units::create("units");
+
+    EXPECT_FALSE(component->isImport());
+    EXPECT_FALSE(units->isImport());
+
+    component->setSourceComponent(imp1, "other_component");
+    units->setSourceUnits(imp1, "other_units");
+
+    EXPECT_TRUE(component->isImport());
+    EXPECT_TRUE(units->isImport());
+    EXPECT_EQ(size_t(1), imp1->componentCount());
+    EXPECT_EQ(size_t(1), imp1->componentCount());
```

Comment:
Did you mean `EXPECT_EQ(size_t(1), imp1->unitsCount());`?

Reply:
2f009081